### PR TITLE
Expose ZIP 32 arbitrary key derivation

### PR DESCRIPTION
**Pull request by str4d**
> **Note**
>This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md).

# Author

- [ ] **Self-review** your own code in GitHub's web interface[^1]
- [ ] Add **automated tests** as appropriate
- [ ] Update the [**manual tests**](../blob/main/docs/testing/manual_testing)[^2] as appropriate
- [ ] Check the **code coverage**[^3] report for the automated tests
- [ ] Update **documentation** as appropriate (e.g [README.md](../blob/main/README.md), [Architecture.md](../blob/main/docs/Architecture.md), etc.)
- [ ] **Run the demo app** and try the changes
- [ ] Pull in the latest changes from the **main** branch and **squash** your commits before assigning a reviewer[^4]

# Reviewer

- [x] Check the code with the [Code Review Guidelines](../blob/main/docs/CODE_REVIEW_GUIDELINES.md) **checklist**
- [ ] Perform an **ad hoc review**[^5]
- [x] Review the **automated tests**
- [ ] Review the **manual tests**
- [x] Review the **documentation**, [README.md](../blob/main/README.md), [Architecture.md](/blob/main/docs/Architecture.md), etc. as appropriate
- [x] **Run the demo app** and try the changes[^6]

[^1]: _Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs._
[^2]: _While we aim for automated testing of the SDK, some aspects require manual testing. If you had to manually test 
something during development of this pull request, write those steps down._
[^3]: _While we are not looking for perfect coverage, the tool can point out potential cases that have been missed. Code coverage can be generated with: `./gradlew check` for Kotlin modules and `./gradlew connectedCheck -PIS_ANDROID_INSTRUMENTATION_TEST_COVERAGE_ENABLED=true` for Android modules._
[^4]: _Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit._
[^5]: _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
[^6]: _While the CI server runs the demo app to look for build failures or crashes, humans running the demo app are 
more likely to notice unexpected log messages, UI inconsistencies, or bad output data. Perform this step last, after verifying the code changes are safe to run locally._

### CODE REVIEW

**Review by HonzaR — COMMENTED** [review]
The proposed changes look good to me. We'll prepare tests to ensure the new APIs return expected values.

**Review by nuttycom — APPROVED** [review]
utACK 968afa46654620269462099d5fc64fb48f6c4724

File: backend-lib/src/main/java/cash/z/ecc/android/sdk/internal/Derivation.kt [hunk 1]
```diff
@@ -38,6 +38,37 @@ interface Derivation {
...
+    /**
+     * Derives a ZIP 32 Arbitrary Key from the given seed at the "wallet level", i.e.
+     * directly from the seed with no ZIP 32 path applied.
+     *
+     * The resulting key will be the same across all networks (Zcash mainnet, Zcash
+     * testnet, OtherCoin mainnet, and so on). You can think of it as a context-specific
+     * seed fingerprint that can be used as (static) key material.
+     *
+     * @param contextString a globally-unique non-empty sequence of at most 252 bytes that
+     *        identifies the desired context.
```
Comment (daira) [inline backend-lib/src/main/java/cash/z/ecc/android/sdk/internal/Derivation.kt:50]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

File: backend-lib/src/main/rust/lib.rs [hunk 2]
```diff
@@ -2003,6 +1837,235 @@ pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_branchIdF
...
+            Err(e) => {
+                return Err(anyhow!(
+                    "Error while deriving viewing key from string input: {}",
+                    e,
+                ));
+            }
+        };
+
+        // Derive the default Unified Address (containing the default Sapling payment
+        // address that older SDKs used).
```
Comment (daira) [inline backend-lib/src/main/rust/lib.rs:1976]: Does that need to be documented at the `RustDerivationTool` layer?

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/tool/DerivationTool.kt [hunk 3]
```diff
@@ -82,6 +82,41 @@ interface DerivationTool {
...
+     * directly from the seed with no ZIP 32 path applied.
+     *
+     * The resulting key will be the same across all networks (Zcash mainnet, Zcash
+     * testnet, OtherCoin mainnet, and so on). You can think of it as a context-specific
+     * seed fingerprint that can be used as (static) key material.
+     *
+     * [ZIP 32 Arbitrary Key]: https://zips.z.cash/zip-0032#specification-arbitrary-key-derivation
+     *
+     * @param contextString a globally-unique non-empty sequence of at most 252 bytes that
+     *        identifies the desired context.
```
Comment (daira) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/tool/DerivationTool.kt:96]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

File: sdk-lib/src/main/java/cash/z/ecc/android/sdk/tool/DerivationTool.kt [hunk 4]
```diff
@@ -82,6 +82,41 @@ interface DerivationTool {
...
+        seed: ByteArray
+    ): ByteArray
+
+    /**
+     * Derives a [ZIP 32 Arbitrary Key] from the given seed at the account level.
+     *
+     * [ZIP 32 Arbitrary Key]: https://zips.z.cash/zip-0032#specification-arbitrary-key-derivation
+     *
+     * @param contextString a globally-unique non-empty sequence of at most 252 bytes that
+     *        identifies the desired context.
```
Comment (daira) [inline sdk-lib/src/main/java/cash/z/ecc/android/sdk/tool/DerivationTool.kt:110]: If failing to meet this requirement can cause a panic, it must also be documented at this layer.

**Review by daira — APPROVED** [review]
utACK with blocking comments.

File: backend-lib/src/main/java/cash/z/ecc/android/sdk/internal/Derivation.kt (hunk 1 above)
Comment (str4d) [reply backend-lib/src/main/java/cash/z/ecc/android/sdk/internal/Derivation.kt:50]: A Rust panic becomes a `RuntimeException` on the Android side, so there's no panic exposed in the public API.

**Review by daira — APPROVED** [review]
utACK. Converting the Rust panic to a Runtime exception resolves my objections, since the preconditions on each function were already documented.

**Review by HonzaR — APPROVED** [review]

**Review by daira — COMMENTED** [review]
Post-hoc ACK

**Pull request merged**